alpha/declcfg/diff.go

```diff
@@ -48,73 +58,101 @@ func (g *DiffGenerator) Run(oldModel, newModel model.Model) (model.Model, error)
 
 	outputModel := model.Model{}
 
-	// Add included packages/channels/bundles from newModel to outputModel.
-	// Code further down handles packages and channels included in outputModel.
-	g.Includer.Logger = g.Logger
-	if err := g.Includer.Run(newModel, outputModel); err != nil {
-		return nil, err
-	}
-
-	if len(oldModel) == 0 {
-		// Heads-only mode.
+	// Prunes old objects from outputModel if they exist.
+	latestPruneFromOutput := func() error {
 
-		// Make shallow copies of packages and channels that are only
-		// filled with channel heads.
-		for _, newPkg := range newModel {
-			// This package may have been created in the include step.
-			outputPkg, pkgIncluded := outputModel[newPkg.Name]
-			if !pkgIncluded {
-				outputPkg = copyPackageNoChannels(newPkg)
-				outputModel[outputPkg.Name] = outputPkg
-			}
-			for _, newCh := range newPkg.Channels {
-				if _, chIncluded := outputPkg.Channels[newCh.Name]; chIncluded {
-					// Head (and other bundles) were added in the include step.
-					continue
-				}
-				outputCh := copyChannelNoBundles(newCh, outputPkg)
-				outputPkg.Channels[outputCh.Name] = outputCh
-				head, err := newCh.Head()
-				if err != nil {
-					return nil, err
-				}
-				outputBundle := copyBundle(head, outputCh, outputPkg)
-				outputModel.AddBundle(*outputBundle)
-			}
-		}
-	} else {
-		// Latest mode.
-
-		// Copy newModel to create an output model by deletion,
-		// which is more succinct than by addition and potentially
-		// more memory efficient.
-		for _, newPkg := range newModel {
-			if _, pkgIncluded := outputModel[newPkg.Name]; pkgIncluded {
-				// The user has specified the state they want this package to have in the diff
-				// via an inclusion entry, so the package created above should not be changed.
-				continue
-			}
-			outputModel[newPkg.Name] = copyPackage(newPkg)
-		}
-
-		// NB(estroz): if a net-new package or channel is published,
-		// this currently adds the entire package. I'm fairly sure
-		// this behavior is ok because the next diff after a new
-		// package is published still has only new data.
 		for _, outputPkg := range outputModel {
 			oldPkg, oldHasPkg := oldModel[outputPkg.Name]
 			if !oldHasPkg {
 				// outputPkg was already copied to outputModel above.
 				continue
 			}
-			if err := diffPackages(oldPkg, outputPkg); err != nil {
-				return nil, err
+			if err := pruneOldFromNewPackage(oldPkg, outputPkg); err != nil {
+				return err
 			}
 			if len(outputPkg.Channels) == 0 {
 				// Remove empty packages.
 				delete(outputModel, outputPkg.Name)
 			}
 		}
+
+		return nil
+	}
+
+	headsOnlyMode := len(oldModel) == 0
+	latestMode := !headsOnlyMode
+	isInclude := len(g.Includer.Packages) != 0
+
+	switch {
```

Comment:
Is a switch statement better than a nested if statement here? Worrying about the default case running or not seems like it would be easy to break on further refactor

Reply:
An if statement would look more confusing, and a switch statement supports future modes more intuitively.